### PR TITLE
Add support for NIST P-256 curve

### PR DESCRIPTION
**Pull request by AlfioEmanueleFresta**
## Changes

* Adds support for NIST P-256, aka secp256r1, using the [p256 crate from RustCrypto](https://github.com/RustCrypto/elliptic-curves/tree/master/p256). This is a pure-Rust implementation.
* As per feedback in previous abandoned PR https://github.com/mcginty/snow/pull/38#issuecomment-435416354, the code is feature gated, as this is a non-standard curve.
* Includes PR #181 which updates curve25519-dalek, which unbreaks the build for me.

## Motivation

This curve is used for the Noise handshake in [CTAP2 Hybrid Transport](https://fidoalliance.org/specs/fido-v2.2-rd-20230321/fido-client-to-authenticator-protocol-v2.2-rd-20230321.html#hybrid-qr-initiated). This is the protocol used for browsers to access FIDO2 credentials (aka Passkeys) on roaming authenticators connected over the internet, such as phones running Android and iOS. 

I'm using snow in my implementation of CTAP2, aimed at bringing Passkeys to the Linux desktop, see [xdg-credentials-portal](https://github.com/AlfioEmanueleFresta/xdg-credentials-portal).

## Tests

All commands below succeed:

```
$ cargo build --all-targets
$ cargo build --all-targets --features p256
$ cargo test --all-targets
$ cargo test --all-targets --features p256
```

### CODE REVIEW

**Comment by AlfioEmanueleFresta** [comment]
Friendy ping on this @mcginty - I would need this change to be upstreamed so we can begin publishing [linux-credentials/libwebauthn](https://github.com/linux-credentials/libwebauthn) to crates.io. TIA!

**Comment by mcginty** [comment]
@AlfioEmanueleFresta Wow, I'm so sorry I dropped the ball on this! Reviewing now.

File: src/constants.rs [hunk 1]
```diff
@@ -4,7 +4,7 @@ pub const TAGLEN: usize = 16;
 
 pub const MAXHASHLEN: usize = 64;
 pub const MAXBLOCKLEN: usize = 128;
-pub const MAXDHLEN: usize = 56;
+pub const MAXDHLEN: usize = 65;
```
Comment (mcginty) [inline src/constants.rs:7]:
```suggestion
// P-256 uncompressed SEC-1 encodings are 65 bytes long, larger than the `MAXDHLEN` in the official Noise spec.
#[cfg(feature = "p256")]
pub const MAXDHLEN: usize = 65;

// Curve448 keys are the largest in the official Noise spec.
#[cfg(not(feature = "p256"))]
pub const MAXDHLEN: usize = 56;
```

File: src/resolvers/default.rs [hunk 2]
```diff
@@ -74,6 +78,14 @@ struct Dh25519 {
     pubkey:  [u8; 32],
 }
 
+/// Wraps p256
+#[cfg(feature = "p256")]
+#[derive(Default)]
+struct P256 {
+    privkey: [u8; 32],
```
Comment (mcginty) [inline src/resolvers/default.rs:85]: Any reason not to store `privkey` as a `p256::SecretKey` here?

File: src/resolvers/default.rs [hunk 3]
```diff
@@ -613,6 +686,27 @@ mod tests {
...
+            "042009fddefeed3b342696b11683b423db8ede2ef5cd66af9b7db2772f7deaf3d\
+                                 f1a69a4648d990ae2a4e5928f156f32e15fa08ba4465df8cb17838dc2afb719d2",
+        )
+        .unwrap();
+        let mut output = [0u8; 32];
+        keypair.dh(&public, &mut output).unwrap();
+        assert_eq!(
+            hex::encode(output),
+            "07505b308650d07e6ead11dd36bbf6f24bf99fc6479649fadd3939faa33ddeb3"
+        );
```
Comment (mcginty) [inline src/resolvers/default.rs:707]: Where did you get these test constants from? It would be good to have a citation to a known-good example to derive trust for this implementation.

**Review by mcginty — CHANGES_REQUESTED** [review]
Thanks for the PR! This is looking pretty good, I just have some small comments/questions.

**Review by mcginty — CHANGES_REQUESTED** [review]
Thanks for the PR! This is looking pretty good, I just have some small comments/questions.

**Comment by complexspaces** [comment]
As a drive-by comment, which might be helpful context: I asked Trevor at RWC one year what his thoughts were on CTAP Hybrid being specified with a non-standard Noise curve, as the designer of Noise, and the answer was that it wasn't a big deal these days and that maybe future versions of Noise would be more curve agnostic anyway.

Tl;dr I think this change is worthwhile for sure :)

**Comment by mcginty** [comment]
@complexspaces thanks for the context! Yeah, generally I'm open to snow supporting "unofficial" general-use Noise modes, as long as the default snow only exposes the officially supported modes and unofficial modifications are feature-gated, just as this PR does.

File: src/resolvers/default.rs (hunk 2 above)
Comment (AlfioEmanueleFresta) [reply src/resolvers/default.rs:85]: I gave this a try again, as I couldn't remember why - it seems it's not trivial:

```
error[E0515]: cannot return reference to temporary value
   --> src/resolvers/default.rs:241:9
    |
241 |         &self.pubkey.to_sec1_bytes()
    |         ^---------------------------
    |         ||
    |         |temporary value created here
    |         returns a reference to data owned by the current function

```

The issue is that the `Dh` trait requires exporting slices:

```rust
pub trait Dh: Send + Sync {
    fn pubkey(&self) -> &[u8];
    fn privkey(&self) -> &[u8];
```

However, `p256::SecretKey` and `p256::PublicKey` only return owned types (eg. `FieldBytes<...>`, or `GenericArray<u8, ...>`).

Pre-serializing the keys to their byte representation allows keeping the trait unchanged.

File: src/resolvers/default.rs (hunk 3 above)
Comment (AlfioEmanueleFresta) [reply src/resolvers/default.rs:707]: I honestly do not recall, so I've replaced these with the test vectors from RFC 5903, and added some comments to this effect.

**Comment by AlfioEmanueleFresta** [comment]
Thanks for the review @mcginty! Addressed your comments.

@complexspaces that's interesting! FWIW, here are some reasons P-256 might have been chosen for the FIDO specs: https://github.com/fido-alliance/discussions/issues/2#issuecomment-2389676305

File: src/resolvers/default.rs (hunk 2 above)
Comment (complexspaces) [reply src/resolvers/default.rs:85]: I also took a quick look and agree this isn't a simple drop-in; the key types available only provide copied-array access to the inner bytes because they are stored as a "generic" field representation.

FWIW I have run into similar issues before with the use of slices for everything in the `Dh` implementations. It was when attempting to add support for ECDH directly in the _ring_ provider because those also don't (at all) provide access to a slice representation of private keys.

**Comment by complexspaces** [comment]
@AlfioEmanueleFresta Yeah it makes a lot of sense why they picked P-256, including the reasons listed in your link. I don't think it was done harmfully to be clear, just out of necessity for what algorithms existing key APIs supported across OSes.

**Review by mcginty — APPROVED** [review]
Looks great, thank you very much for making the change and patience in needing to ping me to review it :).
